src/app/clusters/fan-control-server/fan-control-server.cpp

```diff
@@ -0,0 +1,198 @@
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status);
+
+        // Setting the FanMode value to Off SHALL set the values of PercentSetting, PercentCurrent,
+        // SpeedSetting, SpeedCurrent attributes to 0 (zero):.
+        if (mode == FanModeType::kOff)
+        {
+            status = PercentSetting::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write PercentSetting with error: 0x%02x", status));
+
+            status = PercentCurrent::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write PercentCurrent with error: 0x%02x", status));
+
+            status = SpeedSetting::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write SpeedSetting with error: 0x%02x", status));
+
+            status = SpeedCurrent::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write SpeedCurrent with error: 0x%02x", status));
+        }
+        break;
+    }
+    case PercentSetting::Id: {
+        uint8_t percentSetting;
+        EmberAfStatus status = PercentSetting::Get(attributePath.mEndpointId, &percentSetting);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status);
+
+        // If PercentSetting is set to 0, the server SHALL set the FanMode attribute value to Off.
+        if (percentSetting == 0)
+        {
+            FanModeType currentFanMode;
+            status = FanMode::Get(attributePath.mEndpointId, &currentFanMode);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to get FanMode with error: 0x%02x", status));
+
+            if (currentFanMode != FanModeType::kOff)
+            {
+                status = FanMode::Set(attributePath.mEndpointId, FanModeType::kOff);
+                VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                               ChipLogError(Zcl, "Failed to write FanMode with error: 0x%02x", status));
+            }
+        }
+
+        // Adjust SpeedSetting from a percent value change for PercentSetting
+        // speed = ceil( SpeedMax * (percent * 0.01) )
+        uint8_t speedMax;
+        status = SpeedMax::Get(attributePath.mEndpointId, &speedMax);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status, ChipLogError(Zcl, "Failed to get SpeedMax with error: 0x%02x", status));
+
+        uint8_t currentSpeedSetting;
+        status = SpeedSetting::Get(attributePath.mEndpointId, &currentSpeedSetting);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                       ChipLogError(Zcl, "Failed to get SpeedSetting with error: 0x%02x", status));
```

Comment:
This is not an error condition.  SpeedSetting is an optional attribute.  If the get fails, that means it's not enabled on this endpoint and we should skip the logic below.

Reply:
Added feature bit condition check